File: 프로그래머스/lv2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int n = priorities.length;
        int[] answer = new int[n];
        ArrayDeque<Integer> q = new ArrayDeque<>();
        PriorityQueue<Integer> pq = new PriorityQueue<>((o1, o2) -> o2 - o1);
        for (int i = 0; i < n; i++) {
            pq.add(priorities[i]);
            q.add(i);
        }
        
        for (int i = 0; i < n; i++) {
            int cur = pq.poll();
            for (int j = 0; j < n; j++) {
                if (priorities[q.peekFirst()] != cur) {
                    q.addLast(q.removeFirst());
                } else {
                    answer[q.removeFirst()] = i + 1;
                    break;
                } 
            }
        }
        
        
        return answer[location];
    }
}

File: 프로그래머스/lv2/42587. 프로세스/프로세스.py
from heapq import heappush, heappop
from collections import deque

def solution(priorities, location):
    n = len(priorities)
    answer = [0] * n
    q = deque()
    heap = []
    for i in range(n):
        heappush(heap, -priorities[i])
        q.append(i)
    
    for i in range(n):
        cur = - heappop(heap)
        for _ in range(n):
            if priorities[q[0]] == cur:
                answer[q.popleft()] = i + 1
                break
            else :
                q.append(q.popleft())
    
    return answer[location]